crates/cow-amm/src/maintainers.rs

```diff
@@ -0,0 +1,79 @@
+use {
+    crate::cache::Storage,
+    contracts::ERC20,
+    ethcontract::{futures::future::join_all, Address},
+    ethrpc::Web3,
+    primitive_types::U256,
+    shared::maintenance::Maintaining,
+    std::{
+        collections::{HashMap, HashSet},
+        sync::Arc,
+    },
+    tokio::sync::RwLock,
+};
+
+pub struct EmptyPoolRemoval {
+    storage: Arc<RwLock<Vec<Storage>>>,
+    web3: Web3,
+}
+
+impl EmptyPoolRemoval {
+    pub fn new(storage: Arc<RwLock<Vec<Storage>>>, web3: Web3) -> Self {
+        Self { storage, web3 }
+    }
+}
+
+#[async_trait::async_trait]
+impl Maintaining for EmptyPoolRemoval {
+    async fn run_maintenance(&self) -> anyhow::Result<()> {
+        let mut amms_to_check = HashMap::<Address, HashSet<Address>>::new();
```

Comment:
nit: I don't really see a value in having this complex map, set structure (there won't be any duplicates). Why not simply use a list of `Vec<Amm>`?

Reply:
The idea is to avoid fetching balances of the same AMMs for the same tokens. If I didn't miss anything, the cache might contain the same AMM address with an intersecting list of tokens.

Reply:
How can the cache contain the same AMM twice? A pool can only be deployed once from one factory and pools cannot have the same token listed twice (the smart contract doesn't allow for that).

If we want to be defensive about this, let's implement Hash & Eq on Amm and turn this whole thing into a HashSet (but I think that's over-engineering it)